pkg/kubectl/cmd/drain.go

```diff
@@ -0,0 +1,308 @@
+	options := &DrainOptions{factory: f, out: out}
+
+	cmd := &cobra.Command{
+		Use:     "drain NODE",
+		Short:   "Drain nodes in preparation for maintenance",
+		Long:    drain_long,
+		Example: drain_example,
+		Run: func(cmd *cobra.Command, args []string) {
+			cmdutil.CheckErr(options.SetupDrain(cmd, args))
+			cmdutil.CheckErr(options.RunDrain())
+		},
+	}
+	cmd.Flags().BoolVar(&options.Force, "force", false, "Continue even if there are pods not managed by a ReplicationController.")
+	cmd.Flags().IntVar(&options.GracePeriodSeconds, "grace-period", -1, "Period of time in seconds given to each pod to terminate gracefully. If negative, the default value specified in the pod will be used.")
+	return cmd
+}
+
+// SetupDrain populates some fields from the factory, grabs command line
+// arguments and looks up the node using Builder
+func (o *DrainOptions) SetupDrain(cmd *cobra.Command, args []string) error {
+	var err error
+	if len(args) != 1 {
+		return cmdutil.UsageError(cmd, fmt.Sprintf("USAGE: %s [flags]", cmd.Use))
+	}
+
+	if o.client, err = o.factory.Client(); err != nil {
+		return err
+	}
+
+	o.mapper, o.typer = o.factory.Object()
+
+	cmdNamespace, _, err := o.factory.DefaultNamespace()
+	if err != nil {
+		return err
+	}
+
+	r := o.factory.NewBuilder().
+		NamespaceParam(cmdNamespace).DefaultNamespace().
+		ResourceNames("node", args[0]).
+		Do()
+
+	if err = r.Err(); err != nil {
+		return err
+	}
+
+	return r.Visit(func(info *resource.Info, err error) error {
+		if err != nil {
+			return err
+		}
+		o.nodeInfo = info
+		return nil
+	})
+}
+
+// RunDrain runs the 'drain' command
+func (o *DrainOptions) RunDrain() error {
+	if err := o.RunCordon(true); err != nil {
+		return err
+	}
+
+	pods, err := o.getInterestingPods()
+	if err != nil {
+		return err
+	}
+
+	if err = o.scanForReplicationControllers(pods); err != nil {
+		return err
+	}
+	if err = o.deletePods(pods); err != nil {
+		return err
+	}
+	cmdutil.PrintSuccess(o.mapper, false, o.out, "node", o.nodeInfo.Name, "drained")
+	return nil
+}
+
+// getInterestingPods returns all the pods on the node, but it skips mirror pods
+func (o *DrainOptions) getInterestingPods() ([]api.Pod, error) {
+	pods := []api.Pod{}
+	podList, err := o.client.Pods(api.NamespaceAll).List(unversioned.ListOptions{FieldSelector: unversioned.FieldSelector{Selector: fields.Set{"spec.nodeName": o.nodeInfo.Name}.AsSelector()}})
+	if err != nil {
+		return pods, err
+	}
+	for _, pod := range podList.Items {
+		_, found := pod.ObjectMeta.Annotations[types.ConfigMirrorAnnotationKey]
+		if found {
+			// Skip mirror pod
+			continue
+		}
+
+		pods = append(pods, pod)
+	}
+	return pods, err
+}
+
+// scanForReplicationControllers verifies that all the pods have replication
+// controllers.  If any don't, we abort unless the user used --force.  In that
+// case we just warn.
+func (o *DrainOptions) scanForReplicationControllers(pods []api.Pod) error {
```

Comment:
Rename to scanForReplicatedPods() and have it check for pods managed by ReplicationController or DaemonSet. Then change the comments, documentation, etc. to say that --force is required if there are pods that are not managed by RC or DS.

If you require people to specify --force to drain machines that have DaemonSet pods, then they will basically have to use --force always, which negates the value of having --force.